@kv2019i btw, some build errors:
```
modules/sof/CMakeFiles/modules_sof.dir/__w/sof/sof/sof/src/ipc/ipc3/dai.c.obj.d -o modules/sof/CMakeFiles/modules_sof.dir/__w/sof/sof/sof/src/ipc/ipc3/dai.c.obj -c /__w/sof/sof/sof/src/ipc/ipc3/dai.c
/__w/sof/sof/sof/src/ipc/ipc3/dai.c: In function 'dai_dma_release':
/__w/sof/sof/sof/src/ipc/ipc3/dai.c:314:15: error: 'struct dai_data' has no member named 'chan'
  314 |         if (dd->chan) {
      |               ^~
/__w/sof/sof/sof/src/ipc/ipc3/dai.c:316:44: error: 'struct dai_data' has no member named 'chan'
  316 |                 notifier_unregister(dev, dd->chan, NOTIFIER_ID_DMA_COPY);
      |                                            ^~
/__w/sof/sof/sof/src/ipc/ipc3/dai.c:318:39: error: 'struct dai_data' has no member named 'chan'
  318 |                 dma_release_channel(dd->chan->dma->z_dev, dd->chan->index);
      |                                       ^~
/__w/sof/sof/sof/src/ipc/ipc3/dai.c:318:61: error: 'struct dai_data' has no member named 'chan'
  318 |                 dma_release_channel(dd->chan->dma->z_dev, dd->chan->index);
      |                                                             ^~
/__w/sof/sof/sof/src/ipc/ipc3/dai.c:322:19: error: 'struct dai_data' has no member named 'chan'
  322 |                 dd->chan->dev_data = NULL;
      |                   ^~
/__w/sof/sof/sof/src/ipc/ipc3/dai.c:323:19: error: 'struct dai_data' has no member named 'chan'
  323 |                 dd->chan = NULL;
      |                   ^~
/__w/sof/sof/sof/src/ipc/ipc3/dai.c: In function 'dai_config':
/__w/sof/sof/sof/src/ipc/ipc3/dai.c:355:23: error: 'struct dai_data' has no member named 'chan'
  355 |                 if (dd->chan) {
      |                       ^~
In file included from /__w/sof/sof/zephyr/include/zephyr/logging/log.h:11,
                 from /__w/sof/sof/sof/zephyr/include/sof/trace/trace.h:10,
                 from /__w/sof/sof/sof/zephyr/include/rtos/alloc.h:12,
                 from /__w/sof/sof/sof/zephyr/include/sof/lib/dma.h:23,
                 from /__w/sof/sof/sof/src/include/sof/audio/audio_stream.h:23,
                 from /__w/sof/sof/sof/src/include/sof/audio/buffer.h:11,
                 from /__w/sof/sof/sof/src/include/sof/audio/component.h:19,
                 from /__w/sof/sof/sof/src/include/sof/audio/component_ext.h:19,
                 from /__w/sof/sof/sof/src/ipc/ipc3/dai.c:8:
/__w/sof/sof/sof/src/ipc/ipc3/dai.c:357:37: error: 'struct dai_data' has no member named 'chan'
  357 |                                   dd->chan->index);
```

--- a/src/audio/host-zephyr.c
+++ b/src/audio/host-zephyr.c
@@ -651,7 +651,7 @@ int host_common_trigger(struct host_data *hd, struct comp_dev *dev, int cmd)
 	if (cmd != COMP_TRIGGER_START && hd->copy_type == COMP_COPY_ONE_SHOT)
 		return ret;
 
-	if (hd->chan_index == -EINVAL) {
+	if (hd->chan_index < 0) {
 		comp_err(dev, "no dma channel configured");
 		return -EINVAL;
 	}
@@ -785,7 +785,7 @@ __cold void host_common_free(struct host_data *hd)
 #endif
 
 	/* release DMA channel if not already done by reset */
-	if (hd->chan_index != -EINVAL) {
+	if (hd->chan_index >=  0) {
 		sof_dma_stop(hd->dma, hd->chan_index);
 		sof_dma_release_channel(hd->dma, hd->chan_index);
 		hd->chan_index = -EINVAL;
@@ -1171,7 +1171,7 @@ static int host_position(struct comp_dev *dev,
 
 void host_common_reset(struct host_data *hd, uint16_t state)
 {
-	if (hd->chan_index != -EINVAL) {
+	if (hd->chan_index >= 0) {
 		sof_dma_stop(hd->dma, hd->chan_index);
 		sof_dma_release_channel(hd->dma, hd->chan_index);
 		hd->chan_index = -EINVAL;

--- a/src/ipc/Kconfig
+++ b/src/ipc/Kconfig
@@ -14,6 +14,7 @@ config IPC_MAJOR_3
 
 config IPC_MAJOR_4
 	bool "IPC Major Version 4"
+	depends on ZEPHYR_NATIVE_DRIVERS || LIBRARY
 	help
 	  This is an IPC version used by certain middleware on some IOT
 	  Intel devices. Not for general use.

--- a/src/ipc/ipc4/CMakeLists.txt
+++ b/src/ipc/ipc4/CMakeLists.txt
@@ -2,13 +2,18 @@
 
 # Files common to Zephyr and XTOS
 add_local_sources(sof
-	dai.c
 	handler-user.c
 	handler-kernel.c
 	helper.c
 	logging.c
 	notification.c
 )
+
+# The DAI interface is not implemented in library builds and
+# code depends on Zephyr DAI driver definitions to be available.
+if(NOT CONFIG_LIBRARY)
+	add_local_sources(sof dai.c)
+endif()
 
 is_zephyr(zephyr)
 if(zephyr) ###  Zephyr ###

--- a/src/ipc/ipc4/dai.c
+++ b/src/ipc/ipc4/dai.c
@@ -143,9 +143,7 @@ int ipc_dai_data_config(struct dai_data *dd, struct comp_dev *dev)
 {
 	struct ipc_config_dai *dai = &dd->ipc_config;
 	struct ipc4_copier_module_cfg *copier_cfg = dd->dai_spec_config;
-#ifdef CONFIG_ZEPHYR_NATIVE_DRIVERS
 	struct dai *dai_p = dd->dai;
-#endif
 
 	if (!dai) {
 		comp_err(dev, "no dai!\n");
@@ -174,12 +172,7 @@ int ipc_dai_data_config(struct dai_data *dd, struct comp_dev *dev)
 	case SOF_DAI_INTEL_HDA:
 		break;
 	case SOF_DAI_INTEL_ALH:
-#ifdef CONFIG_ZEPHYR_NATIVE_DRIVERS
 		dd->stream_id = dai_get_stream_id(dai_p, dai->direction);
-#else
-		/* only native Zephyr driver supported */
-		return -EINVAL;
-#endif
 		/* SDW HW FIFO always requires 32bit MSB aligned sample data for
 		 * all formats, such as 8/16/24/32 bits.
 		 */
@@ -243,15 +236,10 @@ void dai_dma_release(struct dai_data *dd, struct comp_dev *dev)
 		 * pause to stop.
 		 * TODO: refine power management when stream is paused
 		 */
-#if CONFIG_ZEPHYR_NATIVE_DRIVERS
 		/* if reset is after pause dma has already been stopped */
 		dma_stop(dd->chan->dma->z_dev, dd->chan->index);
 
 		dma_release_channel(dd->chan->dma->z_dev, dd->chan->index);
-#else
-		dma_stop_legacy(dd->chan);
-		dma_channel_put_legacy(dd->chan);
-#endif
 		dd->chan->dev_data = NULL;
 		dd->chan = NULL;
 	}
@@ -425,7 +413,6 @@ __cold int dai_config(struct dai_data *dd, struct comp_dev *dev,
 			      copier_cfg->gtw_cfg.config_data, size);
 }
 
-#if CONFIG_ZEPHYR_NATIVE_DRIVERS
 int dai_common_position(struct dai_data *dd, struct comp_dev *dev,
 			struct sof_ipc_stream_posn *posn)
 {
@@ -477,51 +464,3 @@ void dai_dma_position_update(struct dai_data *dd, struct comp_dev *dev)
 
 	mailbox_sw_regs_write(dd->slot_info.reg_offset, &slot, sizeof(slot));
 }
-#else
-int dai_common_position(struct dai_data *dd, struct comp_dev *dev,
-			struct sof_ipc_stream_posn *posn)
-{
-	struct dma_chan_status status;
-
-	/* total processed bytes count */
-	posn->dai_posn = dd->total_data_processed;
-
-	platform_dai_wallclock(dev, &dd->wallclock);
-	posn->wallclock = dd->wallclock;
-
-	status.ipc_posn_data = &posn->comp_posn;
-	dma_status_legacy(dd->chan, &status, dev->direction);
-
-	return 0;
-}
-
-int dai_position(struct comp_dev *dev, struct sof_ipc_stream_posn *posn)
-{
-	struct dai_data *dd = comp_get_drvdata(dev);
-
-	return dai_common_position(dd, dev, posn);
-}
-
-void dai_dma_position_update(struct dai_data *dd, struct comp_dev *dev)
-{
-	struct ipc4_llp_reading_slot slot;
-	struct dma_chan_status status;
-	uint32_t llp_data[2];
-
-	if (!dd->slot_info.node_id)
-		return;
-
-	status.ipc_posn_data = llp_data;
-	dma_status_legacy(dd->chan, &status, dev->direction);
-
-	platform_dai_wallclock(dev, &dd->wallclock);
-
-	slot.node_id = dd->slot_info.node_id;
-	slot.reading.llp_l = llp_data[0];
-	slot.reading.llp_u = llp_data[1];
-	slot.reading.wclk_l = (uint32_t)dd->wallclock;
-	slot.reading.wclk_u = (uint32_t)(dd->wallclock >> 32);
-
-	mailbox_sw_regs_write(dd->slot_info.reg_offset, &slot, sizeof(slot));
-}
-#endif

